zio-query/shared/src/main/scala/zio/query/ZQuery.scala

```diff
@@ -121,13 +123,17 @@ final class ZQuery[-R, +E, +A] private (private val step: ZIO[(R, QueryContext),
    * requests may still be applied.
    */
   final def flatMap[R1 <: R, E1 >: E, B](f: A => ZQuery[R1, E1, B]): ZQuery[R1, E1, B] =
-    ZQuery {
-      step.flatMap {
-        case Result.Blocked(br, c) => ZIO.succeedNow(Result.blocked(br, c.mapM(f)))
-        case Result.Done(a)        => f(a).step
-        case Result.Fail(e)        => ZIO.succeedNow(Result.fail(e))
+    new ZQuery(cb =>
+      ZIO.environment[(R1, QueryContext)].flatMap { env =>
```

Comment:
May be worth changing the callback to `(Result[R, E, A] => URIO[R, Any])` so we don't have to access and provide the environment like this.